crates/protocol/src/io.rs

```diff
@@ -0,0 +1,301 @@
+        match self {
+            Some(value) => {
+                true.encode(buffer)?;
+                value.encode(buffer)?;
+            }
+            None => {
+                false.encode(buffer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("unexpected end of input: failed to read value of type `{0}`")]
+    UnexpectedEof(&'static str),
+}
+
+const MAX_LENGTH: usize = 1024 * 1024; // 2^20 elements
+
+pub struct LengthPrefixedVec<'a, P, T>(pub Cow<'a, [T]>, PhantomData<P>)
+where
+    [T]: ToOwned<Owned = Vec<T>>;
+
+impl<'a, P, T> Decoder for LengthPrefixedVec<'a, P, T>
+where
+    T: Decoder,
+    [T]: ToOwned<Owned = Vec<T>>,
+    P: Decoder + TryInto<usize>,
+    P::Error: std::error::Error + Send + Sync + 'static,
+{
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self> {
+        let len = P::decode(buffer)?.try_into()?;
+
+        if len > MAX_LENGTH {
+            bail!(
+                "length {} exceeds maximum allowed length of {}",
+                len,
+                MAX_LENGTH
+            )
+        }
+
+        let vec = iter::repeat_with(|| T::decode(buffer))
+            .take(len)
+            .collect::<anyhow::Result<Vec<T>>>()?;
+
+        Ok(Self(Cow::Owned(vec), PhantomData))
+    }
+}
+
+impl<'a, P, T> Encoder for LengthPrefixedVec<'a, P, T>
+where
+    T: Encoder,
+    [T]: ToOwned<Owned = Vec<T>>,
+    P: Encoder + TryFrom<usize>,
+    P::Error: std::error::Error + Send + Sync + 'static,
+{
+    fn encode(&self, buffer: &mut Vec<u8>) -> anyhow::Result<()> {
+        P::try_from(self.0.len())?.encode(buffer)?;
+
+        self.0.iter().enumerate().for_each(|(index, item)| {
+            item.encode(buffer)
+                .context(format!("failed to encode element at index {}", index))
+                .unwrap();
+        });
+
+        Ok(())
+    }
+}
+
+impl<'a, P, T> From<LengthPrefixedVec<'a, P, T>> for Vec<T>
+where
+    [T]: ToOwned<Owned = Vec<T>>,
+{
+    fn from(x: LengthPrefixedVec<'a, P, T>) -> Self {
+        x.0.into_owned()
+    }
+}
+
+impl<'a, P, T> From<&'a [T]> for LengthPrefixedVec<'a, P, T>
+where
+    [T]: ToOwned<Owned = Vec<T>>,
+{
+    fn from(slice: &'a [T]) -> Self {
+        Self(Cow::Borrowed(slice), PhantomData)
+    }
+}
+
+impl<'a, P, T> From<Vec<T>> for LengthPrefixedVec<'a, P, T>
+where
+    [T]: ToOwned<Owned = Vec<T>>,
+{
+    fn from(vec: Vec<T>) -> Self {
+        Self(Cow::Owned(vec), PhantomData)
+    }
+}
+
+pub type VarIntPrefixedVec<'a, T> = LengthPrefixedVec<'a, VarInt, T>;
+pub type ShortPrefixedVec<'a, T> = LengthPrefixedVec<'a, u16, T>;
+
+pub struct LengthInferredVecU8<'a>(pub Cow<'a, [u8]>);
+
+impl<'a> Decoder for LengthInferredVecU8<'a> {
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self> {
+        let mut vec = Vec::new();
+        buffer.read_to_end(&mut vec)?;
+        Ok(LengthInferredVecU8(Cow::Owned(vec)))
```

Comment:
Universal advice: whenevr `Self` us usable, use it:

```suggestion
        Ok(Self(Cow::Owned(vec)))
```